raft_node.py

```diff
@@ -0,0 +1,595 @@
+        a candidate.
+        """
+        match self._state:
+            case FollowerState(leader) if leader is not None:
+                self._state.leader = new_leader
+            case FollowerState() | CandidateState() | LeaderState():
+                self._state.teardown()
+                self._state = FollowerState(leader=new_leader)
+                run_in_background(self._wait_for_next_heartbeat())
+
+    async def _wait_for_next_heartbeat(self: Self) -> None:
+        """Set up a task that will hopefully be cancelled by an
+        incoming RPC request from a leader within the allotted
+        heartbeat timeout. If it isn't, transition to candidate
+        state. Otherwise, wait for another heartbeat.
+        """
+        loop: asyncio.AbstractEventLoop = asyncio.get_event_loop()
+
+        match self._state:
+            # No way to get into candidate state outside of a heartbeat
+            # timeout, and no way to get into leader state outside of
+            # winning an election
+            case LeaderState() | CandidateState():
+                pass
+
+            case FollowerState(receive_heartbeats_task=recv_heartbeats_task):
+                if recv_heartbeats_task is not None:
+                    # self._print('Heartbeat received. Cancelling old task')
+                    recv_heartbeats_task.cancel()
+
+                wait_for_heartbeat = asyncio.sleep(
+                    self.config.heartbeat_timeout_ms / 1000
+                )
+                self._state.receive_heartbeats_task = loop.create_task(
+                    wait_for_heartbeat
+                )
+
+                try:
+                    await self._state.receive_heartbeats_task
+                    self._state.receive_heartbeats_task = None
+                    self._print("Heartbeat timeout. Becoming a candidate")
+                    self._become_candidate()
+                except CancelledError:
+                    # Heartbeat received
+                    # self._print('Heartbeat task cancelled')
+                    pass
+
+    async def _hold_election(self: Self) -> None:
+        """Send VoteRequests to all other nodes. If in the middle
+        of the election we receive an RPC from a node with a
+        greater or equal term, we'll know another leader was
+        elected and we should instead become a follower.
+
+        If we don't win, we expect another node to win and
+        thus their next heartbeat should cancel the election.
+        If no other node wins, the election will time out and
+        we'll try again.
+        """
+        self._current_term += 1
+        self._voted_for = self.config.raft_server_port
+        self._print(f"Starting election for term {self._current_term}")
+
+        last_log_index, last_log_term = self._log.last_index_and_term()
+        votes: int = 1
+        self._print(
+            "Sending vote requests to "
+            f"{len(self._raft_rpc_clients.values())} clients"
```

Comment:
Do you need to call values() here, or can you just look at the length of the dict?